### PR TITLE
Switch to bootstrapping from 1.27

### PR DESCRIPTION
**Pull request by Mark-Simulacrum**
It's possible the Float trait could be removed from core, but I couldn't tell whether it was intended to be removed or not. @SimonSapin may be able to comment more here; we can presumably also do that in a follow up PR as this one is already quite large.

### CODE REVIEW

File: src/libcore/lib.rs [hunk 1]
```diff
@@ -246,12 +241,9 @@ macro_rules! vector_impl { ($([$f:ident, $($args:tt)*]),*) => { $($f!($($args)*)
 #[path = "../stdsimd/coresimd/mod.rs"]
 #[allow(missing_docs, missing_debug_implementations, dead_code, unused_imports)]
 #[unstable(feature = "stdsimd", issue = "48556")]
-#[cfg(not(stage0))] // allow changes to how stdsimd works in stage0
```
Comment (alexcrichton) [inline src/libcore/lib.rs]: Oh these are actually intended to stay as we permanently, currently, don't want to compile stdsimd in stage (allows for us to tweak it more easily)

File: src/libcore/num/f32.rs [hunk 2]
```diff
@@ -553,7 +552,6 @@ macro_rules! f32_core_methods { () => {
 
 #[lang = "f32"]
 #[cfg(not(test))]
-#[cfg(not(stage0))]
 impl f32 {
     f32_core_methods!();
```
Comment (alexcrichton) [inline src/libcore/num/f32.rs]: I think the idea is that the macro above would be inlined here but we can always do that in a follow-up

**Comment by Mark-Simulacrum** [comment]
I can't really tell exactly what the problem here is, I'll have to dig in deeper. A few other things of note:

Periodically, I get the error below. I think this happens any time I recompile without recompiling the rustc library. Logging shows that the librustc-...rmeta file is rejected due to metadata.

I am somewhat suspecting that a change to metadata encoding/decoding is causing this to occur for sysroot crates, but can't come up for a good reason why that would be the case. I don't immediately see anything in the changelog since last stable to rustc_metadata that could cause this...

I'll dig more into this tomorrow and over the weekend, but I have not come up with a good reason why this would happen.

```
    Checking rustdoc v0.0.0 (file:///home/mark/Build/rust/src/librustdoc)
     Running `/home/mark/Build/rust/build/bootstrap/debug/rustc --crate-name rustdoc librustdoc/lib.rs --error-format json --crate-type lib --emit=dep-info,metadata -C opt-level=3 -C metadata=db2cc8d63ee60236 -C extra-filename=-db2cc8d63ee60236 --out-dir /home/mark/Build/rust/build/x86_64-unknown-linux-gnu/stage0-tools/x86_64-unknown-linux-gnu/release/deps --target x86_64-unknown-linux-gnu -L dependency=/home/mark/Build/rust/build/x86_64-unknown-linux-gnu/stage0-tools/x86_64-unknown-linux-gnu/release/deps -L dependency=/home/mark/Build/rust/build/x86_64-unknown-linux-gnu/stage0-tools/release/deps --extern tempdir=/home/mark/Build/rust/build/x86_64-unknown-linux-gnu/stage0-tools/x86_64-unknown-linux-gnu/release/deps/libtempdir-bbbd9a1d6520a8dc.rmeta --extern pulldown_cmark=/home/mark/Build/rust/build/x86_64-unknown-linux-gnu/stage0-tools/x86_64-unknown-linux-gnu/release/deps/libpulldown_cmark-aabc6d2d653f0cb7.rmeta`
rustc command: "/home/mark/Build/rust/build/x86_64-unknown-linux-gnu/stage0/bin/rustc" "--crate-name" "rustdoc" "librustdoc/lib.rs" "--crate-type" "lib" "--emit=dep-info,metadata" "-C" "opt-level=3" "-C" "metadata=db2cc8d63ee60236" "-C" "extra-filename=-db2cc8d63ee60236" "--out-dir" "/home/mark/Build/rust/build/x86_64-unknown-linux-gnu/stage0-tools/x86_64-unknown-linux-gnu/release/deps" "--target" "x86_64-unknown-linux-gnu" "-L" "dependency=/home/mark/Build/rust/build/x86_64-unknown-linux-gnu/stage0-tools/x86_64-unknown-linux-gnu/release/deps" "-L" "dependency=/home/mark/Build/rust/build/x86_64-unknown-linux-gnu/stage0-tools/release/deps" "--extern" "tempdir=/home/mark/Build/rust/build/x86_64-unknown-linux-gnu/stage0-tools/x86_64-unknown-linux-gnu/release/deps/libtempdir-bbbd9a1d6520a8dc.rmeta" "--extern" "pulldown_cmark=/home/mark/Build/rust/build/x86_64-unknown-linux-gnu/stage0-tools/x86_64-unknown-linux-gnu/release/deps/libpulldown_cmark-aabc6d2d653f0cb7.rmeta" "--cfg" "stage0" "--sysroot" "/home/mark/Build/rust/build/x86_64-unknown-linux-gnu/stage0-sysroot" "-g" "-C" "debug-assertions=y" "-C" "codegen-units=16" "-C" "link-args=-Wl,-rpath,$ORIGIN/../lib" "--color=always" "-Dwarnings"
sysroot: "/home/mark/Build/rust/build/x86_64-unknown-linux-gnu/stage0-sysroot"
libdir: "/home/mark/Build/rust/build/x86_64-unknown-linux-gnu/stage0/lib"
error[E0463]: can't find crate for `rustc`
  --> librustdoc/lib.rs:30:1
   |
30 | extern crate rustc;
   | ^^^^^^^^^^^^^^^^^^^ can't find crate

error: aborting due to previous error

For more information about this error, try `rustc --explain E0463`.
error: Could not compile `rustdoc`.

Caused by:
  process didn't exit successfully: `/home/mark/Build/rust/build/bootstrap/debug/rustc --crate-name rustdoc librustdoc/lib.rs --error-format json --crate-type lib --emit=dep-info,metadata -C opt-level=3 -C metadata=db2cc8d63ee60236 -C extra-filename=-db2cc8d63ee60236 --out-dir /home/mark/Build/rust/build/x86_64-unknown-linux-gnu/stage0-tools/x86_64-unknown-linux-gnu/release/deps --target x86_64-unknown-linux-gnu -L dependency=/home/mark/Build/rust/build/x86_64-unknown-linux-gnu/stage0-tools/x86_64-unknown-linux-gnu/release/deps -L dependency=/home/mark/Build/rust/build/x86_64-unknown-linux-gnu/stage0-tools/release/deps --extern tempdir=/home/mark/Build/rust/build/x86_64-unknown-linux-gnu/stage0-tools/x86_64-unknown-linux-gnu/release/deps/libtempdir-bbbd9a1d6520a8dc.rmeta --extern pulldown_cmark=/home/mark/Build/rust/build/x86_64-unknown-linux-gnu/stage0-tools/x86_64-unknown-linux-gnu/release/deps/libpulldown_cmark-aabc6d2d653f0cb7.rmeta` (exit code: 101)
command did not execute successfully: "/home/mark/Build/rust/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "check" "--target" "x86_64-unknown-linux-gnu" "-j" "16" "-v" "--release" "--manifest-path" "/home/mark/Build/rust/src/tools/rustdoc/Cargo.toml" "--message-format" "json"
```

File: src/Cargo.lock [hunk 3]
```diff
@@ -1565,7 +1565,7 @@ dependencies = [
  "lazy_static 1.0.0 (registry+https://github.com/rust-lang/crates.io-index)",
  "libc 0.2.40 (registry+https://github.com/rust-lang/crates.io-index)",
  "num_cpus 1.8.0 (registry+https://github.com/rust-lang/crates.io-index)",
- "rand 0.4.2 (registry+https://github.com/rust-lang/crates.io-index)",
+ "rand 0.3.22 (registry+https://github.com/rust-lang/crates.io-index)",
```
Comment (Keruspe) [inline src/Cargo.lock]: Is this intended?

**Comment by kennytm** [comment]
Regarding the title... I think you mean "Switch to bootstrapping from 1.**27**"? 😄

**Comment by Mark-Simulacrum** [comment]
Okay, I believe I've tracked down the problem -- caused by https://github.com/rust-lang/rust/pull/49289, but I'm not entirely sure how to fix it. On stable 1.26, running `rustc --crate-name foo foo/src/main.rs --emit=metadata` will emit nothing. On beta 1.27, however, it does emit a libfoo.rmeta file. This is a problem in the Rust repository, as we have `librustc` and `rustc-main` both of which emit the same `libfoo.rmeta` file. This means that when linking the rmeta files into the sysroot, we only consider the latest created metadata in the deps Cargo directory (the rustc-main file), which leads to the "cannot find crate rustc" error I was seeing earlier.

This appears to have been an intentional change; https://github.com/rust-lang/rust/issues/40109 wants the new behavior. I've filed https://github.com/rust-lang/cargo/issues/5524 to track a warning being added to Cargo, and pushed up a commit that should fix it in this PR by renaming the rustc binary to rustc_binary.

The problem that was hit on CI is fixed in the latest pushed commit; I believe something changed in how Cargo reports crate names and renames `-` to `_` so we were failing to detect rustdoc-tool-binary artifact that was generated.

**Comment by Mark-Simulacrum** [comment]
This completed a basic test run locally so I'm going to go ahead an @bors r=alexcrichton

**Comment by Mark-Simulacrum** [comment]
@bors r=alexcrichton

I suspect this might bounce again (please don't rollup); it looks like this compiler needs Copy defined in `no_core` which appears to be relatively prevalent throughout various crates...

**Comment by kennytm** [comment]
The output doesn't appear spurious.

```
[00:04:04] Set({"src/bootstrap"}) not skipped for "test::Bootstrap" -- not in ["src/test/run-pass", "src/test/compile-fail", "src/test/run-pass-fulldeps"]
[00:04:04] Suite("src/test/ui") not skipped for "test::Ui" -- not in ["src/test/run-pass", "src/test/compile-fail", "src/test/run-pass-fulldeps"]
[00:04:04] Suite("src/test/run-pass") not skipped for "test::RunPass" -- not in ["src/test/run-pass", "src/test/compile-fail", "src/test/run-pass-fulldeps"]
[00:04:04] Suite("src/test/compile-fail") not skipped for "test::CompileFail" -- not in ["src/test/run-pass", "src/test/compile-fail", "src/test/run-pass-fulldeps"]
[00:04:04] Suite("src/test/parse-fail") not skipped for "test::ParseFail" -- not in ["src/test/run-pass", "src/test/compile-fail", "src/test/run-pass-fulldeps"]
[00:04:04] Suite("src/test/run-fail") not skipped for "test::RunFail" -- not in ["src/test/run-pass", "src/test/compile-fail", "src/test/run-pass-fulldeps"]
[00:04:05] Suite("src/test/run-pass-valgrind") not skipped for "test::RunPassValgrind" -- not in ["src/test/run-pass", "src/test/compile-fail", "src/test/run-pass-fulldeps"]
[00:04:05] Suite("src/test/mir-opt") not skipped for "test::MirOpt" -- not in ["src/test/run-pass", "src/test/compile-fail", "src/test/run-pass-fulldeps"]
[00:04:05] Suite("src/test/codegen") not skipped for "test::Codegen" -- not in ["src/test/run-pass", "src/test/compile-fail", "src/test/run-pass-fulldeps"]
[00:04:05] Suite("src/test/codegen-units") not skipped for "test::CodegenUnits" -- not in ["src/test/run-pass", "src/test/compile-fail", "src/test/run-pass-fulldeps"]
[00:04:05] Suite("src/test/incremental") not skipped for "test::Incremental" -- not in ["src/test/run-pass", "src/test/compile-fail", "src/test/run-pass-fulldeps"]
[00:04:05] Suite("src/test/debuginfo") not skipped for "test::Debuginfo" -- not in ["src/test/run-pass", "src/test/compile-fail", "src/test/run-pass-fulldeps"]
[00:04:05] Suite("src/test/ui-fulldeps") not skipped for "test::UiFullDeps" -- not in ["src/test/run-pass", "src/test/compile-fail", "src/test/run-pass-fulldeps"]
[00:04:05] Suite("src/test/run-pass-fulldeps") not skipped for "test::RunPassFullDeps" -- not in ["src/test/run-pass", "src/test/compile-fail", "src/test/run-pass-fulldeps"]
[00:04:05] Suite("src/test/run-fail-fulldeps") not skipped for "test::RunFailFullDeps" -- not in ["src/test/run-pass", "src/test/compile-fail", "src/test/run-pass-fulldeps"]
[00:04:05] Suite("src/test/compile-fail-fulldeps") not skipped for "test::CompileFailFullDeps" -- not in ["src/test/run-pass", "src/test/compile-fail", "src/test/run-pass-fulldeps"]
[00:04:05] Suite("src/test/incremental-fulldeps") not skipped for "test::IncrementalFullDeps" -- not in ["src/test/run-pass", "src/test/compile-fail", "src/test/run-pass-fulldeps"]
```

**Comment by Mark-Simulacrum** [comment]
Those lines are present in both successful and unsuccessful logs, though, at least as far as I can tell.

**Comment by kennytm** [comment]
We do have got a log from `i686-mingw-2` now. (`i686-mingw-1` is still the same as [before](https://github.com/rust-lang/rust/pull/50629#issuecomment-388642672)).

```
error[E0405]: cannot find trait `Copy` in this scope
  --> C:\projects\rust\src/rtstartup\rsbegin.rs:61:22
   |
61 |                   impl Copy for $t {}
   |                        ^^^^ not found in this scope
...
66 | /     impl_copy! {
67 | |         usize u8 u16 u32 u64 u128
68 | |         isize i8 i16 i32 i64 i128
69 | |         f32 f64
70 | |         bool char
71 | |     }
   | |_____- in this macro invocation
help: possible candidate is found in another module, you can import it into scope
   |
48 |     use Copy;
   |
error: stability attributes may not be used outside of the standard library
  --> C:\projects\rust\src/rtstartup\rsbegin.rs:60:17
   |
60 |                   #[stable(feature = "rust1", since = "1.0.0")]
   |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
...
66 | /     impl_copy! {
67 | |         usize u8 u16 u32 u64 u128
68 | |         isize i8 i16 i32 i64 i128
69 | |         f32 f64
70 | |         bool char
71 | |     }
   | |_____- in this macro invocation
error: aborting due to 2 previous errors
```

**Comment by Mark-Simulacrum** [comment]
@alexcrichton Any ideas on the curl failure? It looks non-spurious but I also can't tell what could be causing it as I don't think I changed anything curl-sys related. Maybe I'm missing something though...

As an aside, maybe we can just install libcurl-dev in the aux builder?

**Comment by alexcrichton** [comment]
@Mark-Simulacrum this isn't an issue with building libcurl, the real error is:

```
[01:35:38] make[5]: *** write jobserver: Bad file descriptor.  Stop.
```

I have... uh... no idea where that could come from and why it wasn't coming up earlier. AFAIK nothing about that has changed recently

**Comment by Mark-Simulacrum** [comment]
I'm able to reproduce this reliably locally with the docker aux image, but haven't worked on investigating how best to determine the cause. I'm working on reproducing outside docker but still waiting on compilation to finish.

**Comment by Mark-Simulacrum** [comment]
@alexcrichton So I've "fixed" the bug by pre-installing libcurl; I'm not a fan of that fix but maybe we can land this with it for now? What do you think?

**Comment by alexcrichton** [comment]
Hm I would prefer to track this down as we are likely not the only ones who will run into this. I'll try to bottom this out today. If you'd like to land things in the meantime though this is likely related to the Cargo update to updating everything other than that may be feasible

**Comment by alexcrichton** [comment]
Ok I can't really figure out what's going on precisely, but I don't see anything obviously wrong in the strace logs I'm looking (albeit they're millions of lines long).

All in all can the installation of libcurl be backed out and instead this patch applied?

```diff
diff --git a/src/bootstrap/builder.rs b/src/bootstrap/builder.rs
index 17f1922..e4be064 100644
--- a/src/bootstrap/builder.rs
+++ b/src/bootstrap/builder.rs
@@ -586,9 +586,10 @@ impl<'a> Builder<'a> {
 
         // If we were invoked from `make` then that's already got a jobserver
         // set up for us so no need to tell Cargo about jobs all over again.
-        if env::var_os("MAKEFLAGS").is_none() && env::var_os("MFLAGS").is_none() {
-             cargo.arg("-j").arg(self.jobs().to_string());
-        }
+         cargo.arg("-j")
+             .arg(self.jobs().to_string())
+             .env_remove("MAKEFLAGS")
+             .env_remove("MFLAGS");
 
         // FIXME: Temporary fix for https://github.com/rust-lang/cargo/issues/3005
```

The `MAKEFLAGS` env var in effect when bootstrap is running is "obviously wrong" in that it's not actually inheriting anything, so this is just making sure it's not accidentally inherited anywhere else.

My best guess is that this has always been a bug but Cargo's usage of file descriptors was tweaked ever-so-slightly to actually expose this. I don't know what it precisely is though, unfortunately :(

**Comment by Mark-Simulacrum** [comment]
Okay, pushed with the `MAKEFLAGS` patch, leaving in the commit to test on the aux builder on PR Travis. We'll see if it passes.

**Comment by Mark-Simulacrum** [comment]
Aux build passed.

@bors r=alexcrichton

**Pull request merged**